### ISSUE TITLE
CFDP performance

### ISSUE THREAD

**Issue by xpromache** [ASU-cubesat/cfdp-rs#2]
I was testing file transfer with ~30MB files and I see it's pretty slow.

To start with, even without anything going on the cfdp seems to take about 6-7%CPU (in release mode, in debug it's more like 30%). 

Then if I start a transfer, the receiver goes to 100% CPU and it processes incoming PDUs very slowly. The sender and the receiver are connected through UDP. 

Question: have you considered implementing it with async Rust?

**Comment by xpromache** [comment]
Looking a bit through the code I can see a few things:
- the connection between the transactions and transports is done using unbounded channels. That means if the transport is slow, the transaction will send data very fast and it will all accumulate in the channel. Then the transaction will send EOF which will timeout.
-  the UDPTransport does some multiplexing between channels and the UDP socket - in fact it just polls them - I think this polling is responsible for relatively high load when nothing is going on.  I'm not sure the transport needs to actually multiplex the two - I don't see the receive part influencing in any manner the send part. 

If I workaround those two issues I still end up into the transaction.process_pdu - that seems to be fundamentally slow.

to be continued...

**Comment by xpromache** [comment]
Next thing I discovered this sleep here:
https://github.com/ASU-cubesat/cfdp-rs/blob/main/cfdp-core/src/daemon.rs#L664

which I presume is to not have the try_recv go into an infinite loop - but that can be replaced by a simple recv because there is no multiplexing there.

After sorting that out I can see the update_nacks on this line is really slowing it down:
https://github.com/ASU-cubesat/cfdp-rs/blob/main/cfdp-core/src/transaction.rs#L1358

**Comment by mkolopanis** [comment]
Thank you for trying out this crate and your very thorough analysis! I think you have the first real load test of this package and appreciate your feedback!

I have always suspected an asynchronous implementation would be the best choice long term but I personally don't have a lot of experience and wanted to get something up and running. 

I'll implement your suggestions here and try to take a look at update naks to see what can be done better. Or if it could be pushed off to a different part of the transaction to take up less time.

Don't be afraid to make a PR if you'd like to work any changes yourself.

**Comment by xpromache** [comment]
I'll make a pull request to avoid this clone:
https://github.com/ASU-cubesat/cfdp-rs/blob/main/cfdp-core/src/daemon.rs#L1364
which is on the data  reception path and those pdus can be large so I think clone is expensive. The trick is that channel.send returns back the original data if it fails sending it to the channel. 

I will also make a pull request to make the channel between the transaction and the transport bounded. That is to avoid the transaction sending data faster than the transport can deliver.
 

For the nacks I think a good way to handle it is to have an object that deals with reconstructing the file. That one should have an overview of the chunks present in the file and can generate the list of gaps when required (and it's definitely not required at reception of each data chunk).

This is how I implemented in a java implementation I did some time ago:
https://github.com/yamcs/yamcs/blob/master/yamcs-core/src/main/java/org/yamcs/cfdp/DataFile.java

**Comment by mkolopanis** [comment]
About the NAKs, I've been thinking it's probably worth having an enum define which of the NAK modes each transaction should have. Then depending on the mode, a transaction can update the NAK list when necessary (after EOF for delayed, when triggered for async/prompt), but still struggle on when to update NAK lists for immediate mode.

**Comment by xpromache** [comment]
> About the NAKs, I've been thinking it's probably worth having an enum define which of the NAK modes each transaction should have. Then depending on the mode, a transaction can update the NAK list when necessary (after EOF for delayed, when triggered for async/prompt), but still struggle on when to update NAK lists for immediate mode.

I would propose to remove the Transaction.nacks field and replace the update_nacks with get_nacks and call it from the send_nacks function. So you compute it only when needed..

**Comment by xpromache** [comment]
I've been thinking how to eliminate the random sleeps which are in the loop for both send and receive transactions.

We basically have these things to consider in that loop:
- the TX channel to the transport - we have to make this one bounded for the reasons I explained before. If it's bounded the send may block.
- the RX channel from the daemon/transport - reading from it will block when there is no data.
- the timers

The TX/RX channels we can put in a select. The TX has to be put in the select only if there is some data to send (otherwise the select will always return true and we go again in the infinite loop) - so we need a method transaction.has_data_to_send() which looks at all  possible factors (data, nacks, timeouts, eof).

We also need to have the timers in the select. I see two ways: either we keep the current ThreadTimer and make a channel to which that thread timer will add  some message when the timer expires, or simpler - get rid of those thread timers and  compute the remaining times  manually. The computed remaining times can be used then in the select_timeout method when selecting from the RX/TX channels:
https://docs.rs/crossbeam-channel/latest/crossbeam_channel/struct.Select.html#method.select_timeout

Further on, we could have some optimizations for example when the file data is to be sent and there is enough room in the TX channel, we could send multiple PDUs in one go, without going through the select loop. 
Similarly for the receive transactions.

If we do that and solve also the problem of the naks, I think we will be able to achieve very good performance.

**Comment by mkolopanis** [comment]
I wouldn't mind getting rid of the ThreadTimer, just having the transaction compute the times directly is nice and simple. Could even easily support unlimited time if desired easily enough. I like the idea of using the select to eliminate random sleeps.

**Comment by mkolopanis** [comment]
I've gone ahead and remove the FileSizeSensitive enum in favor of a u64 and using a flag from the header to cast to u32 wen necessary.

**Comment by xpromache** [comment]
Here is another idea: what about splitting the Transaction object in two: TxTransaction and RxTransaction (or SendTransaction and ReceiveTransaction). The sender and receiver in CFDP don't have much in common and it will make things clearer - for example it will allow to split the 600 lines process_pdu function in two. 

Looking through existing implementations support this idea:
- The Yamcs implementation (done by me with some colleagues) does that
- I have seen ESA's java implementation they even split the transaction machinery in 4 (TX/RX + ACK/UNACK) 
- NASA cFS also seem to treat the sender and receiver separately: https://github.com/nasa/CF/tree/main/fsw/src


To the question of NAKs:
The title of your  last commit "push nak updating to only after EoF is sent."  is quite confusing because EOF is sent by the Sender whereas the NAKs are computed by the Receiver. If the receiver works in NAK immediate mode, it needs to compute the NAKs before it receives the EOF.
However, I reiterate again that I don't see the point of computing that NAK list on each PDU. It's more efficient to compute it on the fly when you want to send it rather that on each received PDU. 

Another point to consider is the saved segments:
  saved_segments: BTreeMap<u64, u64>

currently it has an entry for each chunk of the file sent/received.

If you have a file of 1GB which is sent in chunks of 1024 bytes, this tree will end up with 1 million entries. And the update_nacks method traverses the tree on the reception of each and every PDU which means for example if you have a link of 10Mbps ~= 1000PDUs/sec, you will have to traverse those 1million entries about 1000 times per second. I don't think it will ever work.

Moreover there is nothing in CFDP which says that the receiver has to agree with the sender in how the file is split into chunks. You could have a receiver that sends a NAK for half a chunk, that's not illegal according to the standard.

Therefore I see no point in keeping that tree like that. First of all the sender doesn't need a tree at all - it sends chunks in order so it only needs to know the offset in the file where is at. When it sends data corresponding to NAKs - well it has in the NAK the data to send.

A more sensible structure for the receiver is to store a list or tree of (start, end) or (start, length) but not corresponding to the received chunks but to the contiguous data segments the receiver accumulates. That means if it receives data in order with no loss it will have one single record in that list. If there is some data loss, it will have more records but gaps usually come in bursts so again it will be just a few records to store. Computing the list of NAKs from that is a breeze.

It will be more work to update that list but assuming that the list remains short, it will also be very efficient. Most of the time the data is just added to the end so practically the end of the last segment is updated - very simple and efficient operation.


BTW: I hope to get this working well with files of 100-200MB and with some luck I will be able to use it on a small experiment on Columbus/ISS later this year. It would be cool, wouldn't it?

**Comment by mkolopanis** [comment]
I think I originally had some idea about wanting unified transaction for simplicity, but admit that separate receive and send transaction could be a simpler approach over all. Probably worth implementing to make it more clear. I was looking at a python implementation that had 4 or so classes I found a little confusing and wanted to simplify, but probably ended up more confusing in the end.

> The title of your last commit "push nak updating to only after EoF is sent." is quite confusing

I think the commit message there is a little confusing. I really just wanted to test something about only calculating NAKs after _receiving_ the EOF, especially since the current implementation really only supports delayed NAK mode. Probably better that each NAK mode of a receiving transaction computes the NAKs when it needs to check them. I think that is what you are suggesting overall and will just take some work to implement.

Thanks for the recommendation on the saved segments. I was thinking as I read through your first part of the message that probably better to save contiguous chunks and that's exactly what you ended up suggesting. 

The experiment you're looking to perform sounds really exciting!

**Comment by xpromache** [comment]
If you want, I can take care of turning the  saved_segments into a list of contiguous chunks. It gets a bit tricky when filling gaps but I already implemented that a few times so it shouldn't get me very long to do it.

**Comment by mkolopanis** [comment]
Sure! I would appreciate that. Any of your ideas you'd like to see implemented feel free to open a PR.  Regarding the bounded channels, would it just require setting the channel between transactions and transports to be `bound(0)` so nothing can sit in the channel?

**Comment by xpromache** [comment]
> Sure! I would appreciate that. Any of your ideas you'd like to see implemented feel free to open a PR.
Ok, I will work to make a contiguous segments PR.


>Regarding the bounded channels, would it just require setting the channel between transactions and transports to be `bound(0)` so nothing can sit in the channel?

I thought it would be that easy but unfortunately not. Making a bounded channel will result in the send blocking - if you are blocked on send you are not able to process incoming messages. Probably you can end up in a deadlock.

I think bounded channels have to work with Select but also that doesn't seem very obvious to implement currently.
Once we split the Sender from the Receiver it will be easier to reason about it. The (un)bounded issue is critical for the Sender, not too much for the Receiver because it doesn't have too much data to send.

**Comment by mkolopanis** [comment]
I'll work on splitting up the transactions into separate structs

**Comment by xpromache** [comment]
> I'll work on splitting up the transactions into separate structs

Please let me know how it goes or if you need some help with that.

I would also like if you don't mind to insert some log::debug! statements in some key moments  - when a new transaction is started, EOF/Finished are received, in case of timeouts. That way it will be easier to monitor what's going on, especially in large transfers.

**Comment by xpromache** [comment]
In fact thinking more about it I will replace "if ... panic" with a simple assert. The rationale is that this is not an external API but an internal thing.

**Comment by xpromache** [comment]
I think this issue can be closed. The cfdp-rs can get easily gigabytes across without so much CPU consumption.

One optimization that could still be done for very high bandwidths is to transfer multiple PDUs at once between the transaction and the transport: that is to use Sender<Vec<PDU>> instead of Sender<PDU> (with the max size of the vector configurable).
But  that's really only necessary when having multi gigabit/s link speeds which is not very common in space communications (it may become more common in the future with the advent of optical links).

In conclusion I suggest we close this issue for now.

**Comment by mkolopanis** [comment]
Sounds good to me. another complication with the `Sender<N>` thing for higher bandwidth communication between the transactions and transports in getting multiple `Permit`s when we move to async mode. But we can cross all these bridges later.

**Comment by xpromache** [comment]
> 

Well I was actually thinking that one permit allows sending multiple PDUs (practically max N - the size of the vector). The PDUs go so fast out that it makes sense to send them in batches. 

On the other hand when  going to >1Gbps link speeds it makes sense to use jumbo frames - so probably could increase the size of the PDU to ~9KB going over standard Ethernet. Copernicus Sentinel missions which make heavy use of CFDP will use PDU sizes of almost 16KB. Going with larger PDUs is more efficient than sending multiple smaller PDUs at once and does not require any modification in cfdp-rs.

**Comment by mkolopanis** [comment]
oh! i see now. I thought of it the wrong way. the channel is for a vec/array/boxed slice of pdus. So each permit can hold that many. It could be a second channel used for fast bandwidth. Interesting.
